blog/_posts/2023-04-20-sbt-plugins-community-repository.md

```diff
@@ -0,0 +1,106 @@
+---
+layout: blog-detail
+post-type: blog
+by: Julien Richard-Foy, Scala Center
+title: "sbt Plugins Community Repository"
+---
+
+The 7th of April 2023, many Scala builds stopped working because an outage affected the sbt community repository.
+In this blog post, we explain what the sbt community repository is and what it stores, we report on the incident
+that happened, and we outline what we can do to avoid such problems in the future.
+
+## The Incident
+
+The 7th of April 2023, many projects of the Scala ecosystem could not anymore resolve
+sbt plugins from the community repository `repo.scala-sbt.org` (examples
+[here](https://github.com/sbt/sbt-dynver/issues/239#issuecomment-1499791434),
+[here](https://github.com/playframework/playframework/issues/11675#issuecomment-1499869916),
+[here](https://github.com/sbt/sbt/issues/7204), and
+[here](https://github.com/spray/sbt-revolver/issues/112)).
+
+## Retrospective
+
+To understand what happened, here is some background information. sbt plugins are packaged
+as modules published to Ivy or Maven repositories. Historically, sbt plugins used to be
+published primarily to the sbt community repository hosted on Bintray. However, in 2021,
+JFrog has [sunsetted the Bintray
+service](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/).
+Since then, JFrog has granted the Scala Center a cloud-hosted Artifactory instance for free.
+As it was explained in the [release notes of sbt
+1.5.1](https://eed3si9n.com/bintray-to-jfrog-artifactory-migration-status-and-sbt-1.5.1),
+this Artifactory instance has been used as a read-only repository to host all the sbt
+plugins that were published to the community repository at that time, but the recommendation
+to plugin authors was to switch to the Central Maven repository.
+
+In practice, the Artifactory instance sponsored by JFrog expires automatically after a
+year, and we have to explicitly ask for renewal. What happened the [7th of
+April](https://github.com/sbt/sbt/issues/7202) is that the instance expired without
+prior notice. Thankfully, Eugene Yokota, the maintainer of sbt, promptly asked JFrog to
+renew the Artifactory instance, which they generously did. After a couple more technical
+steps, everything was back to normal.
+
+## Current State
+
+Thanks to the efforts of Eugene Yokota and the support from the company JFrog, the situation
+is now back to normal: the read-only repository `repo.scala-sbt.org` backed by an Artifactory
+instance works again and all the Scala builds should be able to resolve their plugin
+dependencies as before.
+
+However, as mentioned by Seth Tisue in [a
+comment](https://github.com/sbt/sbt/issues/7202#issuecomment-1500220344), the event
+highlighted a weakness in the Scala ecosystem infrastructure, which is well illustrated by
+this comic:
+
+<p style="text-align: center">
+  <img style="max-width: 385px" src="https://imgs.xkcd.com/comics/dependency.png" alt="Dependency" />
+</p>
```

Comment:
I don't think this comic really serves value in the context of the incident postmortem, in part because it has nothing to do with something thanklessly maintaining a library, and more to do with Artifactory situation and our own incident management.

Reply:
I should've caught this in an earlier draft, but it stands out more as I see the rendering.

Reply:
My thinking in originally considering the comic relevant was that "incident management" is actually often part of "maintaining a library"; they're both under the same umbrella of being responsible for sbt and ensuring its ongoing availability and viability.

Regardless, I definitely don't object to omitting it.

Another reason to omit it is that this blog post is really just an incident report and I think it should be kept short. The end of the post promises "We will let you know as soon as we settle on a plan". In _that_ blog post, perhaps the comic will be more appropriate, I don't know, we'll see.

Reply:
OK, I removed it in [0cc8d29](https://github.com/scala/scala-lang/pull/1492/commits/0cc8d29ba87e47e23c3ba572f952f986d5bc0ae8).